### PR TITLE
Download directory as zip

### LINKED ISSUE #49: [Feature Request] Add support for directory downloading
Like `serve` which sfz draws inspiration from, it would be nice to have a button on each dir for a .zip or .tar.gz of the directory.

### PR DESCRIPTION
**Pull request by whizsid**
Fixes #49 . Please let me know if you want me to make any changes for this PR.

### CODE REVIEW

File: src/server/serve.rs [hunk 1]
```diff
@@ -408,11 +463,12 @@ mod t_server {
...
     #[test]
     fn guess_path_mime() {
-        let mime_type = InnerService::guess_path_mime("file-wthout-extension");
+        let mime_type =
+            InnerService::guess_path_mime("file-wthout-extension", Action::DownloadFile);
         assert_eq!(mime_type, mime::TEXT_PLAIN_UTF_8);
 
         let dir_path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
-        let mime_type = InnerService::guess_path_mime(dir_path);
+        let mime_type = InnerService::guess_path_mime(dir_path, Action::ListDir);
```
Comment (weihanglo) [inline src/server/serve.rs:471]: Maybe add more tests to cover new cases

File: src/server/serve.rs [hunk 2]
```diff
@@ -258,68 +294,89 @@ impl InnerService {
...
-            // Concatenate _modified time_ and _file size_ to
-            // form a (nearly) strong validator.
-            let etag = format!(r#""{}-{}""#, mtime.timestamp(), size)
-                .parse::<ETag>()
-                .unwrap();
-
-            // Validate preconditions of conditional requests.
-            if is_precondition_failed(&req, &etag, mtime) {
-                return Ok(res::precondition_failed(res));
+        match action {
```
Comment (weihanglo) [inline src/server/serve.rs:297]: Nice! I love this pattern matching more readable than before

**Review by weihanglo — COMMENTED** [review]
Overall looks great. Just need some little updates

**Comment by whizsid** [comment]
@weihanglo Changed as you suggested. Can you review it again?

**Review by weihanglo — APPROVED** [review]
LGTM. Thanks for your contribution.

**Pull request merged**